---
title: "Cell-cell_communication analysis"
output: html_document
---

# Introduction to cell-cell communication analysis

*link to powerpoint slides:* A power point presentation will describe the general principles of single-cell 'omics' and basics about the platforms (\~15 - 20 minutes)

------------------------------------------------------------------------

Cells communicate through various mechanisms including chemical signals, physical properties (e.g., mechanotransduction), secreted proteins, and other protein-protein interactions.

![Generated using Biorender.com](images/Untitled%20(7)-01.png){width="586"}

The analysis of cell-cell signaling from single-cell transcriptomic is based on the broad understandings/assumption that [genes encode both ligands and receptors,]{.underline} and sometimes, that activation of a signaling pathway drives and transcriptomic response, both of which can be detected as mRNA expression in individual cells.

------------------------------------------------------------------------

## 0. Setting up the R environment

```{r}
suppressPackageStartupMessages({
  library(CellChat)  
  library(nichenetr)
  library(Seurat)
})

```

**Download pre-processed data for liver from mice treated with TCDD**

```{r}
options(timeout = max(6000, getOption("timeout")))
download.file('https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE148339&format=file&file=GSE148339%5Fliver%2Eintegrated%2Efinal%2ERData%2Egz', './downloads/GSE148339_liver.integrated.final.RData.gz')
system("gunzip ./downloads/GSE148339_liver.integrated.final.RData.gz")
```

**Load the downloaded object**

```{r}
liver_seurat_data = readRDS('./downloads/GSE148339_liver.integrated.final.RData')
liver_seurat_data = UpdateSeuratObject(liver_seurat_data)
```

Next we will take a look at the data to make sure it is loaded properly and that we understand the dataset.

```{r}
DimPlot(liver_seurat_data, group.by = 'celltype', label = TRUE) + NoLegend()
```

For this workshop we will focus on a smaller subset of the data (only endothelial cells and stellate cells).

```{r}
endo_stellate = subset(liver_seurat_data, idents = c("Endothelial Cells", "Stellate Cells", "Macrophages", "Cholangiocytes"))
nonzero_rows <- which(rowSums(as.matrix(endo_stellate@assays$RNA@counts)) != 0)
endo_stellate = endo_stellate[nonzero_rows, ]
endo_stellate@meta.data$celltype = factor(endo_stellate@meta.data$celltype, levels=c('Endothelial Cells', 'Stellate Cells', "Macrophages", "Cholangiocytes"))
```

## 1. Determining *known* cell-cell communication signals

Cell-cell communication analysis tools rely on *a priori* knowledge of interactions. This means that they need to have access to reliable data which represents true and biologically meaningful interactions. Therefore, many of these tools obtain their information from shared source (*e.g.*, KEGG, Reactome, STRING). Nevertheless, there is a lot of heterogeneity in the interactions which are captured across various tools which is explored in further depth by [Dimitrov et al. (2024)](https://www.nature.com/articles/s41467-022-30755-0).

![Similarity between the different resources based on the interactions (Jaccard Index) from Dimitrov, D., TÃ¼rei, D., Garrido-Rodriguez, M. et al. Comparison of methods and resources for cell-cell communication inference from single-cell RNA-Seq data. Nat Commun 13, 3224 (2022). <https://doi.org/10.1038/s41467-022-30755-0>](images/clipboard-747365050.png)

### 1.1. Databases and knowledgebases

As was just highlighted, cell-cell communication analysis is dependent on the underlying database/knowledgebase of existing interactions. Lets take a look at some of the protein-protein interactions which are included in the CellChat database, one of the many commonly use cell-cell communication analysis tools.

```{r}
print(head(CellChatDB.human$interaction))
```

This is an example of how these tools curate interactions for subsequent analyses, identifying the ligand(s), receptor(s), type of signaling, cellular location, and gene symbols. However, its important to note that this database is based on [*known human interactions*]{.underline}.

```{r}
print(paste0(
'Size of human database: ',
nrow(CellChatDB.human$interaction)
))

print(paste0(
'Size of mouse database: ',
nrow(CellChatDB.mouse$interaction)
))

print(paste0(
'Size of zebrafish database: ',
nrow(CellChatDB.zebrafish$interaction)
))
```

We can also look at tools such as NicheNet, which include not only cell-cell interactions, but the subsequent signaling cascade as well. We can load the prior knowledge (PK) as follows:

```{r}
# Load PK
ligand_target_matrix <- readRDS(url("https://zenodo.org/record/7074291/files/ligand_target_matrix_nsga2r_final.rds"))
lr_network <- readRDS(url("https://zenodo.org/record/7074291/files/lr_network_human_21122021.rds"))
```

Next we can examine the database of ligand-receptor interactions. We will filter for TGFB1 as the ligand, just for simplicity.

```{r}
lr_network %>% filter(from == "TGFB1")
```

We can see a structure that is very similar to CellChat. However, NicheNet relies on a second piece of information - the probabilites that certain genes are induced in response to activation of a specific receptor. Here is TGFBR2 for example:

```{r}
ligand_target_matrix['TGFBR2', 1:4]
```

## 2. Steady-state and differential *inter*cellular communication

As with any analysis, whether its in the lab or purely computational, it is important to know what is the right tool to answer your question/test your hypothesis. *Are you looking to characterize cell-cell communication in a tissue without any treatment or perturbation, or are you looking to test the effect of a perturbation on cell-cell communication?* The answers will largely drive which tool you choose, in addition to other questions such as; *how big of a dataset can it handle? what programming language do your prefer to use? etc...*

![Generated using Biorender.com](images/DifferentialCellComm.png){width="338"}

### 2.1. Steady state cell-cell communication

Expression based cell-cell communication analysis tools, including the popular [CellPhoneDB](https://doi.org/10.1038/s41596-020-0292-x) and [CellChat](https://doi.org/10.1038/s41467-021-21246-9) tools, were primarily developed to infer ligand-receptor interactions within cells that are at a steady state. Tools such as CellChat have also developed condition comparison methods. These will be discussed later.

For the exact algorithms which each of these tools use, someone should always review the original publication.

#### 2.1a. CellChat example

In the following example we will explore how CellChat can be used to infer cell-cell communication at steady-state using the mouse liver data downloaded and filtered to only include stellate cells, endothelial cells, macrophages, and cholangiocytes.

**Create cellchat object and choose ligand-receptor database of interest**

```{r}
cellChat <- createCellChat(object = endo_stellate, group.by = "celltype", assay = "RNA")
CellChatDB.use <- CellChatDB.mouse
cellChat@DB <- CellChatDB.use
```

**Remove genes which aren't a ligand or receptor to speed up analysis**

```{r}
# subset the expression data of signaling genes for saving computation cost
cellChat <- subsetData(cellChat) # This step is necessary even if using the whole database
cellChat <- identifyOverExpressedGenes(cellChat)
cellChat <- identifyOverExpressedInteractions(cellChat)
```

**Calculate ligand-receptor pair probabilities**

This is the step with is time consuming in the analysis. Even with the data subset to only four, this can take 10 - 30 minutes.

```{r}
cellChat <- computeCommunProb(cellChat, type = "triMean")
cellChat <- filterCommunication(cellChat, min.cells = 10)
cellchat1 <- computeCommunProbPathway(cellChat)
cellchat1 <- aggregateNet(cellchat1)
```

**Simple visualization**

*Note: We will explore visualizations in further depth in a following section*

```{r}
groupSize <- as.numeric(table(cellchat1@idents))

par(mfrow = c(1,2), xpd=TRUE)
netVisual_circle(cellchat1@net$count, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Number of interactions")
netVisual_circle(cellchat1@net$weight, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")
```

### 2.2. Perturbation of cell-cell communication

#### 2.2a. Enrichment-like approach (NicheNet)

[NicheNet](https://github.com/saeyslab/nichenetr) is different among most cell-cell communication tools in the sense that it does require expression data as input, and it considers downstream effectors such as transcription factors and responses.

Let's consider the question; what ligand-receptor interactions are induced by treatment with chemical X in multiple immune cell types (*e.g.*, T Cells, B Cells, and Macrophages) coming from a source cell type (*e.g.*, hepatocytes). Here we would need to obtain the following [inputs]{.underline}:

1.  **A list of ligands expressed in source/sender cells** and a **list of receptors in the receiving cells**. This list is used to refine the interactions of interest to those that are possible based on gene expression. This can be obtained by finding the intersection of genes in your dataset and the genes in the downloaded knowledgebase as follows:

    ```{r}
    ligands_expressed = intersect(toupper(rownames(endo_stellate)), lr_network$from) 
    receptors_expressed = intersect(toupper(rownames(endo_stellate)), lr_network$to)
      
    potential_interactions <- lr_network %>% 
      filter(from %in% ligands_expressed & to %in% receptors_expressed) %>%
      pull(from) %>% unique()
    ```

2.  **A list of all genes expressed in the receiver cell type**:

    Here we use all genes detected in these cell types by getting the row names of the Seurat object.

    ```{r}
    background_genes = toupper(rownames(endo_stellate))
    ```

3.  **A list of genes which are induced from a treatment or condition**:

    There are several ways to obtain these, but the simplest is to simply perform differential expression between your two cell types of interest. In this example we perform differential expression on the target cell type of interest, the endothelial cells.

    ```{r}

    # Subset seurat object for just stellate cells
    endo = subset(endo_stellate, idents = c("Endothelial Cells"))
    Idents(endo) = "treatment"
                      
    DEAnalysis = FindMarkers(endo, ident.1 = 'TCDD', ident.2 = 'CONTROL', min.pct = 0.5, logfc.treshold = log(2), min.diff.pct = 0.25)
    geneset_oi = toupper(rownames(DEAnalysis %>% filter(avg_log2FC > 0))) # geneset of interest

    print(head(geneset_oi))
    ```

Next we run the analysis of ligand activity and output the top 10 inferred active ligands:

```{r}
ligand_activities <- predict_ligand_activities(
  geneset = geneset_oi, 
  background_expressed_genes = background_genes,
  ligand_target_matrix = ligand_target_matrix,
  potential_ligands = ligands_expressed)

ligand_activities <- ligand_activities %>% 
  arrange(-aupr) %>% 
  mutate(rank = rank(desc(aupr)))

# show top10 ligand activities
head(ligand_activities, n=10)
```

Here we have the top 10 putative active ligands.

#### 2.2b. CellChat

CellChat has similarly incorporated aspects of differential signaling in their analysis package. Here we will run a simple comparison to see what kinds of inputs and outputs are needed, but a much more comprensive tutorial is available from the [CellChat team on Github](https://github.com/jinworks/CellChat).

```{r}

control_seurat <- subset(endo_stellate, subset = treatment == "CONTROL")
treated_seurat <- subset(endo_stellate, subset = treatment == "TCDD")

cellChat_control <- createCellChat(object = control_seurat, group.by = "celltype", assay = "RNA")
cellChat_treated <- createCellChat(object = treated_seurat, group.by = "celltype", assay = "RNA")

object.list <- list(CONTROL = cellChat_control, TCDD = cellChat_treated)

for (cc in names(object.list)){
  cellchat = object.list[[cc]]
  CellChatDB.use = CellChatDB.mouse
  cellchat@DB = CellChatDB.use
  cellchat = subsetData(cellchat)
  cellchat = identifyOverExpressedGenes(cellchat)
  cellchat = identifyOverExpressedInteractions(cellchat)
  cellchat = computeCommunProb(cellchat, type = "triMean")
  cellchat = filterCommunication(cellchat, min.cells = 10)
  cellchat = computeCommunProbPathway(cellchat)
  cellchat = aggregateNet(cellchat)
  object.list[[cc]] = cellchat
  
}

cellchat2 <- mergeCellChat(object.list, add.names = names(object.list))

```

```{r}
gg1 <- rankNet(cellchat2, mode = "comparison", measure = "weight", sources.use = NULL, targets.use = NULL, stacked = T, do.stat = TRUE)
gg2 <- rankNet(cellchat2, mode = "comparison", measure = "weight", sources.use = NULL, targets.use = NULL, stacked = F, do.stat = TRUE)

gg1 + gg2
```

This plot shows how signaling changes with treatment, but as you will see, it considers the overall network does not break it down into individual cell types. Those visualizations will be discussed further in the following section.

## 2. Interpretation and visualization

The diverse tools and their associated tutorials/documentation provide a number of ways to visualize cell-cell communication data, but what can we actually extract from these figures and analyses? Let's explore a few strategies using the CellChat package (*this is by no means comprehensive*).

### 2.1. Broad overview of cell-cell interactions

#### **2.1a. Number of cell-cell interactions *(steady-state)***

We already plotted a network diagram above showing the number and strength of interactions. This is a very high level overview and can often help prioritize subsequent analyses and visualizations. Another way to visualize these global cell-cell communications is through a heatmap:

```{r}
netVisual_heatmap(cellchat1, color.heatmap = "Reds")
```

Here we can see that stellate cells are the primary source (expression of ligands) while endothelial cells arethe primary receiver of signals.

#### 2.1b. Dot plots of active ligand-receptor pairs ***(steady-state)***

We can also ask the question - what ligand-receptor pairs are inferred as active between two cell types. Here we use source #2 (stellate cells) and target #3 (macrophage).

```{r fig.height = 6, fig.width = 4}
netVisual_bubble(cellchat1, sources.use = c(2), targets.use = c(3))
```

We can see, at least with CellChat, that there are several active ligand-receptor pairs. These can be ranked by probability of communication (color scale). For example, here we have the high probability that Interleukin34 (*Il34*) coming from stellate cells is activating Colony Stimulating Factor 1 (*Csf1r*) in macrophage.

### 2.2. Chord diagrams

A chord diagram can provide similar information to the dotplots in a different way. Here the ribbons start at the source and go to the target, allowing similar interactions to be grouped together. We can additionally add the reverse direction to also visualize signals coming from macrophages to stellate cells.

```{r fig.width = 10, fig.height = 10}
netVisual_chord_gene(cellchat1, sources.use = c(2,3), targets.use = c(2,3), legend.pos.x = 15)
```

### 2.3. Sources/Senders vs. Receivers

Some tools, including CellChat, will include options to group classes of ligand-receptor pairs together. For example, in this heatmap of outgoing and incoming interactions we see broad categories such as bone morphogenic proteins and collagens.

```{r fig.width = 10, fig.height = 8}
cellchat1 <- netAnalysis_computeCentrality(cellchat1, slot.name = "netP") 

ht1 <- netAnalysis_signalingRole_heatmap(cellchat1, pattern = "outgoing")
ht2 <- netAnalysis_signalingRole_heatmap(cellchat1, pattern = "incoming")
ht1 + ht2
```

### 2.4. Sankey plots

```{r}
library(NMF)
library(ggalluvial)

selectK(cellchat1, pattern = "outgoing")

nPatterns = 4
cellchat1 <- identifyCommunicationPatterns(cellchat1, pattern = "outgoing", k = nPatterns)

netAnalysis_river(cellchat1, pattern = "outgoing")
```

## 3. Summary
